text/0077-in-memory-pessimistic-locks.md

```diff
@@ -0,0 +1,229 @@
+
+### Region leader transfer
+
+If it is a voluntary leader transfer triggered by PD, we have the chance to transfer the pessimistic locks to the new leader to avoid unexpected transaction failures.
+
+When a peer is going to transfer its leadership, the leader serializes current pessimistic locks into bytes and modifies `valid` to `false`. Then, later `AcquirePessimisticLock` commands will fall back to proposing locks. In this way, we can guarantee the pessimistic locks either exist in the serialized bytes, or are replicated through Raft.
+
+The serialized pessimistic locks will be sent to other peers through a Raft proposal. After this lock migration proposal is committed, the leader will continue the logic of in the current `pre_transfer_leader` method.
+
+By default, a Raft message has a size limit of 1 MiB. We will guarantee that the total size of in-memory pessimistic locks in a single region will not exceed the limit. This will be discussed later.
+
+If the transfer leader message is lost or rejected, we need to revert `valid` to `true`. But it is not possible for the leader to know. So, we have to use a timeout to implement it. That means, we can revert `valid` to `true` if the leader is still not transferred after some period. And if the leader receives the `MsgTransferLeader` response from the follower after the timeout, it should ignore the message and not trigger a leader transfer.
+
+### Region merge
+
+On region merge, the source region can send the pessimistic locks to the target region via `CommitMerge`.
+
+In `propose_normal`, we can first set `valid` to `false` to disable writing pessimistic locks to memory.
+
+In `schedule_merge`, the leader of the source region can serialize current pessimistic locks into bytes and set `valid` to `false`. The serialized pessimistic locks are carried in `CommitMergeRequest`. This means we need an extra field in `CommitMergeRequest`:
+
+```protobuf
+message CommitMergeRequest {
+    // ...
+    bytes pessimistic_locks = 4;
+}
+```
+
+Then, in `exec_commit_merge`, the target region leader deserializes the pessimistic locks from the request and passes it to the `PeerFSM` in `ExecResult::CommitMerge`. The pessimistic locks from the source region are merged to the lock table of the target region in `on_ready_commit_merge`. After it succeeds to merge regions, we increase the pessimistic locks `epoch` and set `valid` to `true`.
+
+If the merge is rolled back, we can set `valid` of the source region back to `true`.
+
+### Region split
+
+Typically, after a region is split into several new regions, the leader of all new regions are still located in the same TiKV. So it will be only memory operations to handle the case.
```

Comment:
What if new region's leader is not on the same TiKV?

Reply:
If the parent of the new region is leader, it will campain to acquire leadership. So I think the possibility of that the new region's leader is not on the same TiKV is roughly like the region leader changes due to timeout.

Reply:
Agree. It's better to clarify in RFC.